src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs

```diff
@@ -0,0 +1,93 @@
+
+namespace Microsoft.Sbom.Api.Executors
+{
+    /// <summary>
+    /// Validates hashes from multiple file locations (on disk and inside SBOM) simulatenously using
+    /// a conncurrent dictionary.
+    /// </summary>
+    public class ConcurrentHashValidator
+    {
+        private readonly FileHashesDictionary fileHashesDictionary;
+
+        public ConcurrentHashValidator(FileHashesDictionary fileHashesDictionary)        
+        {
+            this.fileHashesDictionary = fileHashesDictionary;
+        }
+
+        public (ChannelReader<FileValidationResult> output, ChannelReader<FileValidationResult> errors)
+            Validate(ChannelReader<InternalSBOMFileInfo> fileWithHash)
+        {
+            var output = Channel.CreateUnbounded<FileValidationResult>();
+            var errors = Channel.CreateUnbounded<FileValidationResult>();
+
+            Task.Run(async () =>
+            {
+                await foreach (var fileHash in fileWithHash.ReadAllAsync())
+                {
+                    await Validate(fileHash, output, errors);
+                }
+
+                output.Writer.Complete();
+                errors.Writer.Complete();
+            });
+
+            return (output, errors);
+        }
+
+        private async Task Validate(InternalSBOMFileInfo internalFileInfo, Channel<FileValidationResult> output, Channel<FileValidationResult> errors)
+        {
+            var sha256Checksum = internalFileInfo.Checksum.Where(c => c.Algorithm == AlgorithmName.SHA256).FirstOrDefault();
+            var fileHashes = new FileHashes();
+            fileHashes.SetHash(internalFileInfo.FileLocation, sha256Checksum);
+            FileValidationResult failureResult = null;
+
+            var newValue = fileHashesDictionary.FileHashes.AddOrUpdate(internalFileInfo.Path, fileHashes, (key, oldValue) =>
+            {
+                // This means a file with the same location was already added to the dictionary.
+                if (oldValue.GetHash(internalFileInfo.FileLocation) != null)
+                {
+                    failureResult = new FileValidationResult
+                    {
+                        ErrorType = Entities.ErrorType.AdditionalFile,
+                        Path = internalFileInfo.Path
+                    };
+
+                    return null;
+                }
+
+                oldValue.SetHash(internalFileInfo.FileLocation, sha256Checksum);
+                return oldValue;
+            });
+
+            if (failureResult != null)
+            {
+                await errors.Writer.WriteAsync(failureResult);
+                return;
+            }
+
+            // If we have the files from both locations present in the hash, validate if the hashes match.
+            if (newValue.FileLocation == Sbom.Entities.FileLocation.All)
+            {
+                if (string.Equals(newValue.OnDiskHash.ChecksumValue, newValue.SBOMFileHash.ChecksumValue, StringComparison.InvariantCultureIgnoreCase))
```

Comment:
>InvariantCultureIgnoreCase [](http://example.com/codeflow?start=123&length=26)

Should this be OrdinalComparison?

Reply:
This is a utf 8 string, we convert it from binary to a string representation to serialize to file.